docs/docs/typelevel.md

```diff
@@ -0,0 +1,498 @@
+the following rewritings are performed:
+
+ - `type` is expanded to `inline type def`.
+ - All value parameters get an `inline` modifier.
+ - If the right hand side consists of a conditional, it is put in a spliced block `~{...}`.
+ - The right hand sides of all conditionals are put in type quotes `'[...]`.
+
+We expect that most computed types coming up in practice will use the
+short form. The long form is kept around, to cater for more advanced
+use cases, and to give a precise meaning to the short form.
+
+### Third Step: Query Conditionals
+
+What about defining the inverse of `ToNat`? This should be a function
+that takes a _type_ of peano numbers and produces a compile-time
+integer _value_. We can emulate the effect of such a function with
+implicits but it's a bit roundabout since we would need several
+overloaded methods each taking a different implicit parameter. A more
+direct, closed formulation is possible if we allow implicit searches
+as conditions in spliced code. Example:
+
+    inline def toInt[N <: Nat]: Int = ~{
+      case N =:= Z => '0
+      case N =:= S[type N1] => '(1 + toInt[N1])
+    }
+
+A query conditional is written ` { <cases> }` where each case is of the form
+
+    case <type-pattern> <guard> => <expr>
+
+`<guard>` is either empty or of the form `if <expr>`. Some examples of legal cases are
+
+    case Ord[X] => e1
+    case X <:< List[Y] if n == 0 => e2
+    case _ if n < 0 => e3
+
+Query conditionals may only appear in spliced code. That sets them
+apart from blocks with normal case clauses which define partial
+functions and therefore can only appear in normal code.  Every query
+conditional is evaluated by the macro interpreter. The macro
+interpreter evaluates each case of a query conditional by performing
+an implicit search for the given type pattern. If the search succeeds,
+it continues with evaluating the guard, if there is one, followed by
+the right-hand side, otherwise it continues with the next case.
+
+Type patterns may bind type variables. For instance,
+`N =:= S[type N1]` binds the type variable `N1`. Bound type variables are
+instantiated by the implicit search. They are visible in the rest of
+the case.
+
+In general, a type name in an implicit condition is treated as a bound
+variable if it is prefixed by `type`.
+
+_Aside:_ We should introduce the same convention for types in normal
+patterns. Currently type identifiers in patterns are treated as bound
+if they start with a lower-case letter and as free otherwise. This
+mirrors the convention for names, but is not very visible. Requiring
+`type` prefixes instead would make programs clearer.
+
+In the code above, the macro interpreter would perform an implicit
+search for the type `N =:= S[type N1]` where `N` is the actual type
+passed to `toInt` and `N1` is a type variable bound by the search.
+If the search succeeds, it returns one plus the result of evaluating
+`toInt[N1]`. If the search fails it returns zero.
```

Comment:
There is no `case _ => '0` brunch. Should it be something like "If the search fails it results in compilation failure"?